Docs: add terraform setup for Organization EC2 Discovery (#68679)
gravitational/teleport: docs/pages/enroll-resources/auto-discovery/servers/ec2-discovery/ec2-discovery-organization-terraform.mdx

---
title: Terraform EC2 Auto-Discovery Configuration for AWS Organizations
sidebar_label: Terraform Setup for Organizations
sidebar_position: 5
description: How to configure Teleport EC2 auto-discovery with Terraform for an AWS Organization
page_type: how-to
tags:
 - how-to
 - zero-trust
 - infrastructure-identity
 - aws
---

> This guide shows you how to use Terraform to configure Teleport and AWS to automatically enroll EC2 instances in your cluster.

Reviewer: ```suggestion
This guide shows you how to use Terraform to configure Teleport and AWS to automatically enroll EC2 instances from your AWS organization into your Teleport cluster.
```

    }
  }

  # Each child account in the organization must have an IAM role with this name.
  # This role is assumed by the discovery service to enroll resources from that account.
  # The required trust relationship and permissions for this role can be found in the module outputs and documentation.
  aws_child_account_iam_role_name = "teleport-organization-discovery-child-account-role"

  # Discover EC2 instances using matching rules.
  # Accepts "*" to discover across all enabled regions.
  # The module adds account:ListRegions to the IAM policy automatically when "*" is used.
  aws_matchers = [
    {
      types = ["ec2"]
      # EC2 discovery supports a wildcard to find instances in all regions.
      regions = ["*"]
      tags = {
        env = ["prod"]
      }
    }
  ]
}
```
</TabItem>

{/*
TODO(gavin): add instructions for the scenario where a self-hosted Teleport cluster cannot be reached on the public internet.
In this case, they need to use a different IAM credential source because OIDC federation will not work. 
The discovery team has not yet come to a consensus on how we want to approach this use-case, so it is currently out of scope.

> */}

Reviewer: @marcoandredinis @GavinFrazar I thought we support/tested this use-case already with our tf modules? The mode where it doesn't create OIDC integration and requires discovery service to have an existing IAM role. Is this note obsolete?

Author: I added the docs for users to create the required IAM roles for Discovery and Auth Services.

Now I'm wondering whether I should just create those Roles or whether users should create them manually.

I'll keep them as is for now, but open to feedback for a new iteration on the module.

## Prerequisites

(!docs/pages/includes/edition-prereqs-tabs.mdx!)

- EC2 instances running Ubuntu/Debian/RHEL/Amazon Linux 2/Amazon Linux 2023 and SSM Agent version 3.1 or greater
- (!docs/pages/includes/tctl.mdx!)
- Access to the management account or a member account that is a delegated administrator.
- AWS Terraform provider configured with credentials from the management account or a member account that is a delegated administrator.
- Teleport Terraform provider configured using one of the methods described in [Using the Teleport Terraform Provider](../../../../configuration/terraform-provider/terraform-provider.mdx).

<Admonition type="note">

All EC2 instances added to the Teleport cluster by the Discovery Service must
include the `AmazonSSMManagedInstanceCore` IAM policy in order to receive
commands from the Discovery Service. For a list of permissions included in the
policy, see the [AWS
documentation](https://docs.aws.amazon.com/aws-managed-policy/latest/reference/AmazonSSMManagedInstanceCore.html).

</Admonition>

> <Admonition type="note">

Reviewer 2: ```suggestion
```
The second Admonition consists entirely of the `<details>` box, so I think we can fold it into the first one.

The Teleport Discovery Service queries the AWS API to list matching Accounts, and EC2 instances in each account.
For each EC2 instance that it discovers, the Discovery Service uses AWS Systems Manager (SSM) to install Teleport on the instance and join it to the cluster as a Teleport-protected server.

## Prerequisites

(!docs/pages/includes/edition-prereqs-tabs.mdx!)

- EC2 instances running Ubuntu/Debian/RHEL/Amazon Linux 2/Amazon Linux 2023 and SSM Agent version 3.1 or greater
- (!docs/pages/includes/tctl.mdx!)
- Access to the management account or a member account that is a delegated administrator.

<Admonition type="note">

All EC2 instances added to the Teleport cluster by the Discovery Service must
include the `AmazonSSMManagedInstanceCore` IAM policy in order to receive
commands from the Discovery Service. For a list of permissions included in the
policy, see the [AWS
documentation](https://docs.aws.amazon.com/aws-managed-policy/latest/reference/AmazonSSMManagedInstanceCore.html).

</Admonition>

> ## Install Teleport to run the Discovery Service

Reviewer 2: We've been standardizing on, in the body of a how-to guide, using the `## Step x/y` format for H2 headers. This helps the reader understand how far along they are in the instructions.

In this case, there are two H2s for instructions, but there's not a way to offset them from H2s that are outside the main instructional flow, like "Updating module configuration" or "Troubleshooting". It's also unclear whether the guide intends for the user to read "Install Teleport to run the Discovery Service", _then_ "Use the teleport-discovery-aws Terraform module", or if they are alternatives to one another. The current structure leaves a lot for the reader to figure out.

I would either use a single sequence of `## Step x/y` headers or split this into two guides with clear signposting between them, whichever one is clearer for a reader.

Author: Yeah, I can see how this can be confusing.
I was trying to follow existing sibling guides which also use this format.

I will merge both sequences

<Admonition type="tip">

If you already have a running Discovery Service instance, then assign its `discovery_group` to <Var name="discovery-group-name" /> and proceed to [use the `teleport-discovery-aws` Terraform module](#use-the-teleport-discovery-aws-terraform-module).

All Teleport Cloud clusters run the Discovery Service for you, so Teleport Cloud subscribers can also skip all of these installation steps.

</Admonition>

### Step 1/3. Install Teleport

If you plan on running the Discovery Service on the same host already running
another Teleport service (Auth or Proxy, for example), then you can skip this step.

Install Teleport on a host instance that will run the Discovery Service:

(!docs/pages/includes/install-linux.mdx!)

### Step 2/3. Configure the Discovery Service

If you are running the Discovery Service on its own host, the service requires a
valid invite token to connect to the cluster. Generate one by running the
following command against your Teleport Auth Service:

```code
$ tctl tokens add --type=discovery
```

> Save the generated token in `/tmp/token` on the Node (EC2 instance) that will

Reviewer 2: We don't use the capitalized "Node" concept in the docs anymore (besides some legacy instances we haven't removed yet), since it's ambiguous as to whether it means "a Teleport Agent instance" or "a Teleport-protected server".

run the Discovery Service.

(!docs/pages/includes/discovery/discovery-group.mdx!)

Assign <Var name="teleport.example.com:443" /> to the host and port of the Teleport Proxy Service in your cluster,
and <Var name="discovery-group-name" /> to a name that identifies a group of resources that you will enroll:

```yaml
# teleport.yaml
version: v3
teleport:
  join_params:
    # token_name can be a literal token string or a path to a file.
    # File path is preferable to avoid including a secret in the config file.
    token_name: "/tmp/token"
    method: token
  proxy_server: "<Var name="teleport.example.com:443" />"
auth_service:
  enabled: false
proxy_service:
  enabled: false
ssh_service:
  enabled: false
discovery_service:
  enabled: true
  discovery_group: "<Var name="discovery-group-name" />"
```

### Step 3/3. Start Teleport

(!docs/pages/includes/aws-credentials.mdx service="the Discovery Service"!)

(!docs/pages/includes/start-teleport.mdx service="the Discovery Service"!)

## Use the `teleport-discovery-aws` Terraform module

### Step 1/6. Configure AWS Terraform provider

> Configure the [AWS Terraform provider](https://registry.terraform.io/providers/hashicorp/aws/latest/docs) and AWS IAM permissions for Terraform to manage AWS resources.

Reviewer 2: I would try not to include links in a "Step" section of a how-to guide unless there's a clear objective to complete when following the link. This way, we keep the reader's attention on the guide. 

Instead of relying on a link, is there a minimal AWS Terraform provider config we can include as an example? As an alternative, we could make configuring the AWS Terraform provider a prerequisite. Right now, this part of the guide is pretty vague.

Author: > I would try not to include links in a "Step" section of a how-to guide unless there's a clear objective to complete when following the link. This way, we keep the reader's attention on the guide.

We have two important things in this step:
- tell the user they need to configure AWS Terraform provider
- let them know which permissions are required for the identity that will be assigned to the provider

So, this is something they really have to follow in order to complete the guide.

Option A
Move this to a prerequisite, with a link to https://registry.terraform.io/providers/hashicorp/aws/latest/docs
Valid, but then where do we show the required permissions?
Adding them to the prerequisite doesn't seem feasible.
Not showing them is also not really an option.

Option B
Provide a minimal AWS Terraform provider config
Customers have so many ways of configuring AWS that I'm not sure providing one example would work for a majority of users.
Looking at the docs, we could replicate their first example on how to configure the AWS Provider
However that is using credentials directly in the terraform file, which honestly is not at all recommended. And something I would rather not advertise on our docs.

https://registry.terraform.io/providers/hashicorp/aws/latest/docs#provider-configuration

> Right now, this part of the guide is pretty vague.

Is it? Users reading this guide are already aware that this is for configuring AWS resource enrollment using terraform.

I would imagine that in most scenarios, users already have Terraform configured for AWS in some pipeline.

I'm open to suggestions tho.
If you still think that adding an example for AWS terraform provider configuration is worth it, I can copy one from their official docs.
I don't like that idea because we are pretty much copying things around and saying how to configure product X from product Y, and that eventually leads to out of date docs and/or outdated recommended approaches.

Reviewer 2: I don't see instructions for configuring the AWS Terraform provider, only a set of permissions required. Since users likely already have a configuration present, and already know how to apply the permissions, I think this is more appropriate for the Prerequisites than a step within the guide: it's not a set of instructions to follow, but rather a set of preconditions that we expect users to have before they can proceed.

Author: Yeah, we don't have instructions to guide users in configuring the AWS Terraform provider.
But we tell them to.

My concern with moving this to the Prerequisites is that it might be too confusing if we show the list of required permissions there, and scare them 😅 

What do you think?

        "iam:ListPolicyTags",
        "iam:ListPolicyVersions",
        "iam:TagPolicy",
        "iam:UntagPolicy"
      ],
      "Resource": "*"
    },
    {
      "Sid": "ManageRolePolicyAttachments",
      "Effect": "Allow",
      "Action": [
        "iam:AttachRolePolicy",
        "iam:DetachRolePolicy",
        "iam:ListAttachedRolePolicies"
      ],
      "Resource": "*"
    },
    {
      "Sid": "ManageOidcProvider",
      "Effect": "Allow",
      "Action": [
        "iam:CreateOpenIDConnectProvider",
        "iam:DeleteOpenIDConnectProvider",
        "iam:GetOpenIDConnectProvider",
        "iam:ListOpenIDConnectProviders",
        "iam:TagOpenIDConnectProvider",
        "iam:UntagOpenIDConnectProvider",
        "iam:UpdateOpenIDConnectProviderThumbprint"
      ],
      "Resource": "*"
    },
    {
      "Sid": "Organization",
      "Effect": "Allow",
      "Action": [
        "organizations:ListAccounts",
        "organizations:ListRoots",
        "organizations:ListAWSServiceAccessForOrganization",
        "organizations:DescribeOrganization"
      ],
      "Resource": "*"
    }
  ]
}
```

</details>

### Step 2/6. Configure Teleport Terraform provider

> There are several ways to configure the Teleport Terraform provider depending on how you intend to run Terraform, for example in CI, Spacelift, or some other remote environment, but for a quick start, see the [Local Demo](../../../../configuration/terraform-provider/local.mdx) guide.

Reviewer 2: We should make this a prerequisite instead of having the reader interrupt their progress to follow another guide. We could also move this to a "Further reading" section.

Author: See answer below

The local demo guide walks through configuring the Teleport Terraform provider with local Teleport credentials, which is typically as simple as logging in with `tsh` and running a `tctl` command from the same shell that you use to run `terraform` commands:

```code
$ tsh login
$ eval "$(tctl terraform env)"
```

> If you are running Terraform in a remote environment, such as a cloud VM, an on-prem server, or CI/CD pipelines, refer to [Using the Teleport Terraform Provider](../../../../configuration/terraform-provider/terraform-provider.mdx) to find the appropriate guide for your use case.

Reviewer 2: Same with this paragraph. Instead of having the reader open a several browser tabs and figure out where to go, we should make this a prerequisite so, by the time they're ready for this guide, they can keep it open the entire time.

Author: We are not asking them to go read another guide here.
We tell them how to configure things quickly (ie `eval "$(tctl terraform env)"`) but also link to other guides in case their setup is more complete (eg, has CI/CD).

Isn't what we are doing here pretty much what you've asked for the AWS Terraform provider configuration: "is there a minimal ... Terraform provider config we can include as an example?" ?
https://github.com/gravitational/teleport/pull/68679#discussion_r3623513058

Reviewer 2: > We are not asking them to go read another guide here.

The first paragraph says, "for a quick start, see the Local Demo guide", explicitly asking them to navigate away from the page. For other situations, Step 5 asks the user to choose a guide to read.

I think this guide would save the reader some friction by making it a prerequisite to follow one of the Teleport Terraform provider setup guides.

Author: Can you please take another look? Thank you

### Step 3/6. Configure the Terraform module inputs

Add the `teleport-discovery-aws` module to your Terraform configuration.

<Tabs>

> <TabItem label="cloud">

Reviewer 2: ```suggestion
<TabItem label="Teleport Cloud">
```

```hcl
module "aws_discovery" {
  source = "terraform.releases.teleport.dev/teleport/discovery/aws"
  version = "~> (=cloud.major_version=).0"

  # Required inputs:

>   # Assign <Var name="example.teleport.sh:443" /> to your Teleport cluster's proxy public address in host:port form.

Reviewer 2: ```suggestion
  # Assign <Var name="example.teleport.sh:443" /> to the public address of the Teleport Proxy Service in your cluster in host:port form.
```
Using non-standard names for Teleport components adds friction, especially when a Teleport cluster has a several kinds of proxy server.

  teleport_proxy_public_addr = "<Var name="example.teleport.sh:443" />"
  # teleport_discovery_group_name must match the discovery group name in your Discovery Service config file.
  # Teleport Cloud clusters run the Discovery Service in the group name "cloud-discovery-group".
  # Do not modify this input unless you intend to run your own Discovery Service.
  teleport_discovery_group_name = "cloud-discovery-group"

>   # Enable AWS Organization wide discovery.

Reviewer 2: The name of the product is AWS Organizations. What is the role of "wide" here?

Author: Removed it.
What I was trying to convey is already implied by the "AWS Organizations" name

    }
  }

  # Each child account in the organization must have an IAM role with this name.
  # This role is assumed by the Discovery Service to enroll resources from that account.
  # The required trust relationship and permissions for this role can be found in the module outputs and documentation.
  aws_child_account_iam_role_name = "teleport-organization-discovery-child-account-role"

  # Optional inputs:
  # Apply the additional AWS tag "origin=example" to all AWS resources created by this module
  apply_aws_tags = { origin = "example" }
  # Apply the additional Teleport label "origin=example" to all Teleport resources created by this module
  apply_teleport_resource_labels = { origin = "example" }

  # Discover EC2 instances using matching rules.
  # Accepts "*" to discover across all enabled regions.
  # The module adds account:ListRegions to the IAM policy automatically when "*" is used.
  aws_matchers = [
    {
      types = ["ec2"]
      # EC2 discovery supports a wildcard to find instances in all regions.
      regions = ["*"]
      tags = {
        env = ["prod"]
      }
    }
  ]
}
```
</TabItem>

> <TabItem label="self-hosted">

Reviewer 2: ```suggestion
<TabItem label="Self-hosted">
```

```hcl
module "aws_discovery" {
  source = "terraform.releases.teleport.dev/teleport/discovery/aws"
  version = "~> (=teleport.major_version=).0"

  # Required inputs:
  # Edit this input to the host and port of the Teleport Proxy Service in your cluster
  teleport_proxy_public_addr    = "<Var name="teleport.example.com:443" />"
  # Edit this input to match the discovery group name in your Discovery Service config file
  teleport_discovery_group_name = "<Var name="discovery-group-name" />"

  # Enable AWS Organization wide discovery.
  aws_organization_discovery = {
    organizational_units = {
      # The wildcard "*" matches all organizational units.
      # You can also specify a list of organizational unit IDs to match (eg, "ou-1a2b-rg12qq2b")
      include = ["*"]

      # You can also specify a list of organizational unit IDs to exclude from discovery.
      # exclude = ["ou-1a2b-rg12qq2b"]
      # Excluded organizational units will not be discovered, even if they are included in the "include" list.
    }
  }

  # Each child account in the organization must have an IAM role with this name.

>   # This role is assumed by the discovery service to enroll resources from that account.

Reviewer 2: ```suggestion
  # This role is assumed by the Discovery Service to enroll resources from that account.
```
For consistent capitalization

  # Discover EC2 instances using matching rules.
  # Accepts "*" to discover across all enabled regions.
  # The module adds account:ListRegions to the IAM policy automatically when "*" is used.
  aws_matchers = [
    {
      types = ["ec2"]
      # EC2 discovery supports a wildcard to find instances in all regions.
      regions = ["*"]
      tags = {
        env = ["prod"]
      }
    }
  ]
}
```

<Admonition type="note">

By default, the `teleport-discovery-aws` Terraform module configures an AWS OIDC integration for discovery and enrollment of EC2 instances.
However, this requires the Teleport Proxy Service to be reachable on the public internet.

If your cluster is running in a private network and the Proxy Service is not reachable on the public internet, you can disable the OIDC integration and assign the necessary IAM permissions for discovering and enrolling EC2 instances.

You can disable the AWS OIDC integration by setting the following:
```hcl
  discovery_service_iam_credential_source = {
    use_oidc_integration = false
  }
  # This role must be created manually and be assigned to the Discovery Service.

>   aws_organization_discovery_iam_principal_arn = "arn:aws:iam::<root-account-id>:role/teleport-organization-account-enumeration"

Author: The DiscoveryService must use an IAM role which belongs to the root or delegated administrator account anyway.
So, it will end up being the same account id as the rest of the guide, which we are referring to as `root-account-id`

```hcl
aws_discovery = {
  # ...
  "teleport_organization_account_enumeration_iam_role_template" = {
    "policy" = <<-EOT
    {
      "Version": "2012-10-17",
      "Statement": [
        {
          "Effect": "Allow",
          "Action": [
            "organizations:ListRoots",
            "organizations:ListChildren",
            "organizations:ListAccountsForParent"
          ],
          "Resource": "*"
        },
        {
          "Effect": "Allow",
          "Action": "sts:AssumeRole",
          "Resource": "arn:aws:iam::*:role/teleport-organization-discovery-child-account-role",
          "Condition": {
            "StringEquals": {
              "aws:ResourceOrgID": "<aws-organization-id>"
            }
          }
        }
      ]
    }
    EOT
    "role_name" = "teleport-organization-account-enumeration"
  }
  "teleport_organization_join_validation_iam_role_template" = {
    "policy" = <<-EOT
    {
      "Version": "2012-10-17",
      "Statement": [
        {
          "Effect": "Allow",
          "Action": "organizations:DescribeAccount",
          "Resource": "*"
        }
      ]
    }
    EOT
    "role_name" = "teleport-organization-join-validation"
  }
}
```

> The Discovery Service attached IAM role must have the policy defined in `teleport_organization_account_enumeration_iam_role_template` and the Auth Service attached IAM role must have the policy defined in `teleport_organization_join_validation_iam_role_template`.

Reviewer 2: ```suggestion
The IAM role attached to the Discovery Service must have the policy defined in `teleport_organization_account_enumeration_iam_role_template`. The role attached to the Auth Service must have the policy defined in `teleport_organization_join_validation_iam_role_template`.
```
Grammar
